### PR TITLE
[Linux] Collect dmesg, /var/log/messages and journals at virtual devices, firmwares tests failure

### PR DESCRIPTION
**Pull request by keirazhang**
When there is virtual firmware or devices testing failures, dmesg and linux logs can provide useful information for triage. That's why we need changes in this fix, which include:
1. Collect dmesg output, and extract call trace from it
2. Collect /var/log/messages and `journalctl -b` output
3. Collect guest and VM info both for PVRDMA client VM and PVRDMA server VM at `pvrdma_network_device_ops` failure.

### CODE REVIEW

File: linux/utils/collect_dmesg.yml [hunk 1]
```diff
@@ -4,25 +4,43 @@
...
-    fetch_file_dst_path: "{{ current_test_log_folder }}/"
-
-- name: "Set facts of dmesg output file at localhost"
-  ansible.builtin.set_fact:
-    dmesg_output_file_local: "{{ fetch_file_local_path }}"
+- name: "Save dmesg output to local file in test case log folder"
   when:
-    - fetch_file_local_path is defined
-    - fetch_file_local_path
+    - dmesg_output_file_name is defined
```
Comment (Tomorrow9) [inline linux/utils/collect_dmesg.yml:27]: So this parameter 'dmesg_output_file_name' does not have a default value, if it's not defined, then will not save output to a file.

File: linux/utils/collect_dmesg.yml [hunk 2]
```diff
@@ -4,25 +4,47 @@
...
 #   Collect dmesg and save output to a local file
 # Parameters:
-#   dmesg_opts: The dmesg command options
-#   dmesg_output_file_src: Save dmesg output file on guest OS
+#   dmesg_options: the dmesg command options
+#   dmesg_output_file_name (optional): the file name for saving dmesg output in test case log folder
+#   dmesg_ignore_errors (optional): ignore errors when executing dmesg. Default is false.
 # Return:
-#   dmesg_output_file_local: The dmesg output file at localhost
+#   dmesg_output_file_path: the file path to saved dmesg output at localhost
```
Comment (123lzxm) [inline linux/utils/collect_dmesg.yml:11]: saved-->save

File: linux/utils/collect_dmesg.yml [hunk 3]
```diff
@@ -4,25 +4,47 @@
...
-    - fetch_file_local_path
+    - dmesg_cmd_result.rc is defined
+    - dmesg_cmd_result.rc == 0
+    - dmesg_cmd_result.stdout is defined
+    - dmesg_cmd_result.stdout
+  block:
+    # Set a default dmesg output file with execution timestamp
+    - name: "Set default file name for saving the dmesg command output"
+      ansible.builtin.set_fact:
+        dmesg_output_file_name: "dmesg_{{ lookup('pipe', 'date +%Y%m%d%H%M%S') }}.log"
```
Comment (123lzxm) [inline linux/utils/collect_dmesg.yml:35]: It can use "current_test_timestamp" directly.

File: linux/setup/linux_test_rescue.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#
+- name: "Collect dmesg at test case failure"
+  include_tasks: ../utils/collect_dmesg.yml
+  vars:
+    dmesg_ignore_errors: true
+    dmesg_output_file_name: "dmesg_at_test_failure.log"
+
+- name: "Detect errors in dmesg output"
+  when:
+    - dmesg_output_file_path is defined
```
Comment (123lzxm) [inline linux/setup/linux_test_rescue.yml:15]: "dmesg_output_file_path" has been initialized in above task. No need to check if it's defined.

File: linux/cpu_hot_add_basic/cpu_hot_add_basic.yml [hunk 5]
```diff
@@ -15,40 +15,57 @@
   tasks:
     - name: "Test case block"
       block:
-        - include_tasks: ../../common/skip_test_case.yml
+        # Except for VM guest_id like 'fedora64Guest', 'freebsd.*Guest' not supporting vCPU hot-add,
+        # the other guest OS compatible with ansible-vsphere-gos-validation support vCPU hot-add
+        - name: "Skip test case '{{ ansible_play_name }}' due to VM not support vCPU hot-add"
```
Comment (123lzxm) [inline linux/cpu_hot_add_basic/cpu_hot_add_basic.yml:20]: "VM" --> "guest OS" is better?

File: linux/cpu_hot_add_basic/cpu_hot_add_basic.yml [hunk 6]
```diff
@@ -15,40 +15,57 @@
...
             cpu_hotadd_results: []
 
-        - include_tasks: cpu_hot_add_prepare.yml
+        - name: "Reconfigure VM with initial vCPU number and vCPU hot-add enabled"
+          include_tasks: cpu_hot_add_prepare.yml
           vars:
             enable_vm_cpu_hotadd: true
-        - include_tasks: cpu_hot_add_validate.yml
+
+        - name: "Check CPU and cores number in guest OS after vCPU hot-add"
```
Comment (123lzxm) [inline linux/cpu_hot_add_basic/cpu_hot_add_basic.yml:58]: CPU--> vCPU

File: linux/cpu_multicores_per_socket/cpu_multicores_per_socket.yml [hunk 7]
```diff
@@ -15,33 +15,46 @@
...
             cpu_multicores_per_socket_results: []
 
-        - include_tasks: ../cpu_hot_add_basic/cpu_hot_add_prepare.yml
+        - name: "Reconfigure VM with initial vCPU and cores number"
+          include_tasks: ../cpu_hot_add_basic/cpu_hot_add_prepare.yml
           vars:
             enable_vm_cpu_hotadd: false
-        - include_tasks: cpu_multicores_validate.yml
+
+        - name: "Check CPU number and cores in guest OS for VM with multiple vCPU cores"
```
Comment (123lzxm) [inline linux/cpu_multicores_per_socket/cpu_multicores_per_socket.yml:47]: CPU --> vCPU

File: linux/device_list/device_list.yml [hunk 8]
```diff
@@ -8,38 +8,48 @@
...
           vars:
+            create_test_log_folder: true
             skip_test_no_vmtools: true
 
-        - include_tasks: ../utils/device_connectable_enable.yml
+        - name: "Enable VM device config for connection and disconnection"
+          include_tasks: ../utils/device_connectable_enable.yml
 
-        - include_tasks: ../utils/get_vmware_toolbox_cmd_path.yml
+        - name: "Get VMware Tool toolbox command path"
```
Comment (123lzxm) [inline linux/device_list/device_list.yml:28]: Tool --> Tools

File: linux/utils/collect_dmesg.yml (hunk 3 above)
Comment (keirazhang) [reply linux/utils/collect_dmesg.yml:35]: No, it can't use here. Because this task may be executed more than once in a test case.

File: linux/cpu_hot_add_basic/cpu_hot_add_basic.yml (hunk 6 above)
Comment (keirazhang) [reply linux/cpu_hot_add_basic/cpu_hot_add_basic.yml:58]: The first CPU is from guest OS perspective, which doens't know it is virtual or physical.

File: linux/cpu_hot_add_basic/cpu_hot_add_basic.yml (hunk 5 above)
Comment (keirazhang) [reply linux/cpu_hot_add_basic/cpu_hot_add_basic.yml:20]: No. It's because the VM's guest id doesn't support cpu hot-add, no matter what the real guest OS is.

File: linux/utils/collect_dmesg.yml (hunk 2 above)
Comment (keirazhang) [reply linux/utils/collect_dmesg.yml]: Updated. Thanks!

File: linux/secureboot_enable_disable/secureboot_enable_disable.yml [hunk 9]
```diff
@@ -35,28 +40,33 @@
...
+        - name: "Check VM image is approved in vmware.log when secure boot is enabled"
+          include_tasks: ../../common/vm_wait_log_msg.yml
           vars:
             vm_wait_log_name: "vmware.log"
             vm_wait_log_msg: "Image APPROVED"
             vm_wait_log_retries: 1
             vm_wait_log_ignore_errors: true
 
-        - include_tasks: ../utils/get_system_firmware.yml
+        - name: "Get system firmware in guest OS"
```
Comment (123lzxm) [inline linux/secureboot_enable_disable/secureboot_enable_disable.yml:68]: No need to get the firmware here, and condition on line 76 can also be removed, since if firmware==bios, the test case has already been skipped at the beginning.

File: linux/secureboot_enable_disable/secureboot_enable_disable.yml [hunk 10]
```diff
@@ -66,19 +76,24 @@
...
+          include_tasks: ../../common/vm_set_power_state.yml
           vars:
             vm_power_state_set: 'powered-on'
 
-        - include_tasks: ../../common/update_inventory.yml
+        - name: "Update inventory"
+          include_tasks: ../../common/update_inventory.yml
 
-        - include_tasks: ../utils/get_system_firmware.yml
+        - name: "Get system firmware in guest OS"
```
Comment (123lzxm) [inline linux/secureboot_enable_disable/secureboot_enable_disable.yml:95]: Same comment as above.

File: linux/network_device_ops/pvrdma_client_network_device_ops.yml [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+          ansible.builtin.copy:
+            dest: "{{ current_test_log_folder }}/ib_devinfo.json"
+            content: "{{ guest_rdma_device_info }}"
+
+    - name: "Collect client VM information at test failure"
+      include_tasks: ../../common/test_rescue.yml
+      vars:
+        vm_dir_name: "{{ pvrdma_client_vm_dir_name }}"
+        vm_vmx_path: "{{ pvrdma_client_vm_vmx_path }}"
+      when:
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/network_device_ops/pvrdma_client_network_device_ops.yml:137]: When the two vars are not set, it will skip test rescue tasks.

File: linux/memory_hot_add_basic/memory_hot_add_basic.yml [hunk 12]
```diff
@@ -14,38 +14,64 @@
...
+        - name: "Test case is blocked due to empty memory size list"
+          include_tasks: ../../common/skip_test_case.yml
           vars:
-            skip_msg: "Test case '{{ ansible_play_name }}' is blocked because memory hotadd test value list is empty"
+            skip_msg: "Test case '{{ ansible_play_name }}' is blocked because memory hot-add test value list is empty"
             skip_reason: "Blocked"
           when: memory_hotadd_size_list | length == 0
 
-        - include_tasks: ../../common/skip_test_case.yml
+        # From guest OS, Debian for 32bit and FreeBSD for 32bit and 64bit don't support memory hot-add.
```
Comment (Tomorrow9) [inline linux/memory_hot_add_basic/memory_hot_add_basic.yml:47]: There is `vm_guest_id is match('freebsd.*')` above, can this part also be moved as above to match guest ID?

File: linux/network_device_ops/pvrdma_network_device_ops.yml [hunk 13]
```diff
@@ -12,11 +12,36 @@
...
+        - name: "Create test log folders for server VM and client VM"
+          include_tasks: ../../common/create_directory.yml
+          vars:
+            dir_path: "{{ item }}"
+            dir_mode: "0755"
+          with_items:
+            - "{{ pvrdma_server_test_log_folder }}"
+            - "{{ pvrdma_client_test_log_folder }}"
+
+        - name: "Set fact of test log folder for server VM"
```
Comment (Tomorrow9) [inline linux/network_device_ops/pvrdma_network_device_ops.yml:41]: - name: "Set current test log folder to the server VM test log folder"

File: linux/network_device_ops/pvrdma_client_network_device_ops.yml [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+# Perform network device operations on client VM
+#
+- name: "Test block on PVRDMA client VM"
+  block:
+    - name: "Set facts to perform network device operations on client VM"
+      ansible.builtin.set_fact:
+        vm_name: "{{ pvrdma_client_vm_name }}"
+        current_test_log_folder: "{{ pvrdma_client_test_log_folder }}"
+        pvrdma_client_vm_ready: false
+        pvrdma_client_test_success: false
```
Comment (Tomorrow9) [inline linux/network_device_ops/pvrdma_client_network_device_ops.yml:13]: This parameter `pvrdma_client_test_success` can be replaced by setting to 'fail' in 'rescue' part directly.

File: linux/memory_hot_add_basic/memory_hot_add_basic.yml (hunk 12 above)
Comment (keirazhang) [reply linux/memory_hot_add_basic/memory_hot_add_basic.yml:47]: `guest_os_ansible_distribution` is set in `test_setup.yml` for existing VM, so it can't be moved above. I will file another PR to move out `get_linux_system_info.yml` from `test_setup.yml`, and then here can be improved.

File: linux/cpu_hot_add_basic/cpu_hot_add_basic.yml [hunk 15]
```diff
@@ -15,40 +15,57 @@
...
     - name: "Test case block"
       block:
-        - include_tasks: ../../common/skip_test_case.yml
+        # Except for VM guest_id like 'fedora64Guest', 'freebsd.*Guest' not supporting vCPU hot-add,
+        # the other guest OS compatible with ansible-vsphere-gos-validation support vCPU hot-add
+        - name: "Skip test case '{{ ansible_play_name }}' due to VM not support vCPU hot-add"
+          include_tasks: ../../common/skip_test_case.yml
           vars:
-            skip_msg: "Skip test case '{{ ansible_play_name }}' because vCPU hotadd is not supported by {{ guest_id }}"
+            skip_msg: "Skip test case '{{ ansible_play_name }}' because vCPU hot-add is not supported by guest id {{ vm_guest_id }}"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/cpu_hot_add_basic/cpu_hot_add_basic.yml]: updated.

File: linux/cpu_multicores_per_socket/cpu_multicores_per_socket.yml [hunk 16]
```diff
@@ -15,33 +15,46 @@
...
     - name: "Test case block"
       block:
-        - name: Set fact of the initial VM vCPU number
+        - name: "Set facts of the VM initial vCPU and cores number"
           ansible.builtin.set_fact:
             initial_cpu_num: 2
             initial_cores_num: 2
 
-        - include_tasks: ../cpu_hot_add_basic/generate_cpu_hot_add_list.yml
+        - name: "Generate vCPU number list for hot-add"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/cpu_multicores_per_socket/cpu_multicores_per_socket.yml]: fixed

File: linux/memory_hot_add_basic/memory_hot_add_basic.yml [hunk 17]
```diff
@@ -14,38 +14,64 @@
   tasks:
     - name: "Test case block"
       block:
+        # Except for VM guest_id like 'freebsd.*Guest' not supporting memory hot-add,
+        # the other guest OS compatible with ansible-vsphere-gos-validation support memory hot-add
+        - name: "Skip test case '{{ ansible_play_name }}' due to VM not support memory hot-add"
+          include_tasks: ../../common/skip_test_case.yml
+          vars:
+            skip_msg: "Skip test case '{{ ansible_play_name }}' because memory hot-add is not supported by guest id {{ vm_guest_id }}"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/memory_hot_add_basic/memory_hot_add_basic.yml]: updated

File: linux/network_device_ops/pvrdma_client_network_device_ops.yml (hunk 14 above)
Comment (keirazhang) [reply linux/network_device_ops/pvrdma_client_network_device_ops.yml]: update.d

File: linux/network_device_ops/pvrdma_client_network_device_ops.yml [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+      include_tasks: ../../common/vm_set_power_state.yml
+      vars:
+        vm_power_state_set: 'powered-off'
+
+    - name: "Remove the client VM"
+      include_tasks: ../../common/vm_remove.yml
+
+    - name: "Set fact of test success on client VM"
+      ansible.builtin.set_fact:
+        pvrdma_client_test_success: true
```
Comment (Tomorrow9) [inline linux/network_device_ops/pvrdma_client_network_device_ops.yml:111]: I mean if there is task failed in this block, then tasks in below `rescue` will be executed, at the end can add a `fail` task, then will go to the outside `rescue`, so this parameter 'pvrdma_client_test_success' is no need to be set here and be validated outside in `pvrdma_network_device_ops.yml` line 107.

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks!

**Comment by Tomorrow9** [comment]
Will this structure be workable?
```
test case block
    - test prepare
    - test on client VM block
       client rescue (only handle client VM)
    - test on server VM
       server rescue (only handle server VM)
test case rescue (only handle test case related)
```

**Comment by keirazhang** [comment]
> Will this structure be workable?
> 
> ```
> test case block
>     - test prepare
>     - test on client VM block
>        client rescue (only handle client VM)
>     - test on server VM
>        server rescue (only handle server VM)
> test case rescue (only handle test case related)
> ```

I am afraid it can't. Operations on client VM are not completely same with server VM. And test on client VM happens between hot-add and hot-remove network adapter on server VM. They can't be completely separated.

**Comment by keirazhang** [comment]
@Tomorrow9 @123lzxm do you have any comments about new changes?

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks!
